04-HTML5-APIs/app.js

```diff
@@ -0,0 +1,174 @@
+
+    localStorage.setItem(key, value)
+    if (localStorage.getItem(key) === null)
+        throw new Error("There was an error when saving");
+
+}
+
+const clearLocal = (key) => {
+
+    if (localStorage.getItem(key))
+        localStorage.removeItem(key)
+    else
+        throw new Error("The item does not exist");
+
+
+}
+
+
+
+let request = window.indexedDB.open("DB1");
+let db;
+
+request.onupgradeneeded = function(event) {
+    db = request.result;
+    db.createObjectStore("savedText", { autoIncrement: false });
+};
+
+request.onsuccess = function(event) {
+    db = request.result;
+};
+
+request.onerror = function(event) {
+    Error(event.target.errorCode);
+};
+
+const saveDb = (key, value) => {
+    let tx = db.transaction(key, "readwrite");
+    let store = tx.objectStore(key);
+
+    store.put(value, key);
+}
+
+const clearDb = (key) => {
+    let tx = db.transaction(key, "readwrite");
+    let store = tx.objectStore(key);
+
+    store.clear();
+}
+
+const saveTextArea = (event) => {
+    saveLocal("savedText", textArea.value);
+    saveDb("savedText", textArea.value);
+}
+
+const clearTextArea = (event) => {
+    clearLocal("savedText");
+    clearDb("savedText");
+}
+
+saveBtn.addEventListener("click", saveTextArea);
+clearBtn.addEventListener("click", clearTextArea);
+
+///////////////Exercise 2////////////////
+
+const dropZone = document.getElementById("dropZone");
+const dropData = document.getElementById("txtFileData");
+
+const overDropZone = (event) => {
+    event.preventDefault();
+    console.log("File is in the drop zone");
+}
+
+const enterDropZone = (event) => {
+    event.preventDefault();
+    console.log("File entered in the drop zone");
+    dropZone.classList.add("dropZoneReady");
+}
+
+const leaveDropZone = (event) => {
+    event.preventDefault();
+    console.log("File has left the drop zone");
+    dropZone.classList.remove("dropZoneReady");
+}
+
+const handleFiles = (file) => {
+
+}
+
+const readFile = (file) => {
+
+    let reader = new FileReader();
+    reader.readAsText(file);
+    if (file.type == "text/plain") {
+        reader.onload = function(event) {
+            dropData.textContent += event.target.result;
+        }
+    } else {
+        dropData.textContent += "Not a .txt file";
+    }
+}
+
+const dropTxt = (event) => {
+    event.preventDefault();
+    dropData.textContent = "";
+    let file = event.dataTransfer.files[0];
+    readFile(file);
+    dropZone.classList.remove("dropZoneReady");
+}
+
+
+
+
+dropZone.addEventListener("dragenter", enterDropZone);
+dropZone.addEventListener("dragleave", leaveDropZone);
+dropZone.addEventListener("dragover", overDropZone)
+dropZone.addEventListener("drop", dropTxt);
+
+///////////////Exercise 3////////////////
+
+let wsUri = "wss://echo.websocket.org/";
+let output;
+
+function init() {
+    output = document.getElementById("webSocketOutput");
+    testWebSocket();
+}
+
+const testWebSocket = () => {
+    websocket = new WebSocket(wsUri);
```

Comment:
:x: Una forma facil de evitar esta variable global es simplemente asignando to funcion al attributo y usar this cuando precises usar el websocker. Revisate la leccion anterior de oop para entender el this y preguntame si algo no quedo claro. Pero lo usarias algo asi.

websocket.onmessage = onMessage;

// lines of code

function onMessage(evt) {
    writeToScreen('<span style="color: blue;">RESPONSE: ' + evt.data+'</span>');
    this.close();
}

Reply:
Si, es evidentemente una mala costumbre que tengo el tema de las variables locales, ya lo voy a corregir.